### PR TITLE
Add support for Core Authorization's User Roles and Permissions

### PR DESCRIPTION
**Pull request by jakbutler**
### Changes
- Added a new `RolesEntity` for providing access to the [Role management API](https://auth0.com/docs/api/management/v2#!/Roles) endpoints. 
- Added supporting `Role`, `Permission`, `RolesPage`, and `PermissionPage` value objects, with respective deserializers.
- Added a `RolesFilter` for use in the [get_roles](https://auth0.com/docs/api/management/v2#!/Roles/get_roles) call.
- Added a simple `PageFilter` for specifying filter objects that only contain pagination options, e.g. `per_page` and `include_totals`.
- Updated `UserEntity` to include [User Permission](https://auth0.com/docs/api/management/v2#!/Users/get_permissions) and [User Role](https://auth0.com/docs/api/management/v2#!/Users/get_roles) endpoints. 

### References
Feature Request #192 

### Testing
- [x] This change adds test coverage
- [x] This change has been tested on the latest version of the platform/language

### Checklist
- [x] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [x] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
- [x] All existing and new tests complete without errors

### CODE REVIEW

**Comment by lbalmaceda** [comment]
@jakbutler thanks for taking care of this. I've added a review task to the backlog and I hope I can start with it the next week.

**Comment by jakbutler** [comment]
Apologies, I'd forgotten to also update the `README.md` to mention the Roles entity; that's been added now.

**Comment by martijnblankestijn** [comment]
Terrific work @jakbutler ! I was just researching how we could use the Roles API and found out it was missing. A very nice surprise to see this PR.  I can't wait until a new release is being made with this feature added. I

File: src/main/java/com/auth0/client/mgmt/UsersEntity.java [hunk 1]
```diff
@@ -350,4 +355,213 @@ public Request deleteMultifactorProvider(String userId, String provider) {
...
+    }
+
+    /**
+     * Get the roles associated with a user.
+     * See https://auth0.com/docs/api/management/v2#!/Users/get_roles
+     *
+     * @param userId the role id
+     * @return a Request to execute
+     */
+    public Request<List<Role>> listRoles(String userId) {
```
Comment (lbalmaceda) [inline src/main/java/com/auth0/client/mgmt/UsersEntity.java]: kill this

File: src/main/java/com/auth0/client/mgmt/UsersEntity.java [hunk 2]
```diff
@@ -350,4 +355,213 @@ public Request deleteMultifactorProvider(String userId, String provider) {
...
+            .addPathSegments("roles").build().toString();
+        CustomRequest<List<Role>> request = new CustomRequest<>(client, url, "GET", new TypeReference<List<Role>>() {
+        });
+        request.addHeader("Authorization", "Bearer " + apiToken);
+        return request;
+    }
+
+    /**
+     * Get the roles associated with a user.
+     * See https://auth0.com/docs/api/management/v2#!/Users/get_roles
```
Comment (lbalmaceda) [inline src/main/java/com/auth0/client/mgmt/UsersEntity.java]: missing scopes (in new line) `read:users` `read:roles`. 
Also URL should be https://auth0.com/docs/api/management/v2#!/Users/get_user_roles

File: src/main/java/com/auth0/client/mgmt/UsersEntity.java [hunk 3]
```diff
@@ -350,4 +355,213 @@ public Request deleteMultifactorProvider(String userId, String provider) {
...
+        CustomRequest<RolesPage> request = new CustomRequest<>(client, url, "GET", new TypeReference<RolesPage>() {
+        });
+        request.addHeader("Authorization", "Bearer " + apiToken);
+        return request;
+    }
+
+
+    /**
+     * Remove roles from a user. A token with update:users is needed.
+     * See https://auth0.com/docs/api/management/v2#!/Users/delete_roles
```
Comment (lbalmaceda) [inline src/main/java/com/auth0/client/mgmt/UsersEntity.java]: should be https://auth0.com/docs/api/management/v2#!/Users/delete_user_roles

File: src/main/java/com/auth0/client/mgmt/UsersEntity.java [hunk 4]
```diff
@@ -350,4 +355,213 @@ public Request deleteMultifactorProvider(String userId, String provider) {
...
+        String url = builder.build().toString();
+        CustomRequest<RolesPage> request = new CustomRequest<>(client, url, "GET", new TypeReference<RolesPage>() {
+        });
+        request.addHeader("Authorization", "Bearer " + apiToken);
+        return request;
+    }
+
+
+    /**
+     * Remove roles from a user. A token with update:users is needed.
```
Comment (lbalmaceda) [inline src/main/java/com/auth0/client/mgmt/UsersEntity.java]: new line

File: src/main/java/com/auth0/client/mgmt/UsersEntity.java [hunk 5]
```diff
@@ -350,4 +355,213 @@ public Request deleteMultifactorProvider(String userId, String provider) {
...
+            .toString();
+        VoidRequest request = new VoidRequest(this.client, url, "DELETE");
+        request.setBody(body);
+        request.addHeader("Authorization", "Bearer " + apiToken);
+        return request;
+    }
+
+    /**
+     * Assign roles to a user. A token with update:users is needed.
+     * See https://auth0.com/docs/api/management/v2#!/Users/post_roles
```
Comment (lbalmaceda) [inline src/main/java/com/auth0/client/mgmt/UsersEntity.java]: should be https://auth0.com/docs/api/management/v2#!/Users/post_user_roles

File: src/main/java/com/auth0/client/mgmt/UsersEntity.java [hunk 6]
```diff
@@ -350,4 +355,213 @@ public Request deleteMultifactorProvider(String userId, String provider) {
...
+            .build()
+            .toString();
+        VoidRequest request = new VoidRequest(this.client, url, "DELETE");
+        request.setBody(body);
+        request.addHeader("Authorization", "Bearer " + apiToken);
+        return request;
+    }
+
+    /**
+     * Assign roles to a user. A token with update:users is needed.
```
Comment (lbalmaceda) [inline src/main/java/com/auth0/client/mgmt/UsersEntity.java]: new line

File: src/main/java/com/auth0/client/mgmt/filter/PageFilter.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
+package com.auth0.client.mgmt.filter;
+
+/**
+ * Class used to filter the results received when calling the Grants endpoint. Related to the {@link com.auth0.client.mgmt.RolesEntity} entity.
```
Comment (lbalmaceda) [inline src/main/java/com/auth0/client/mgmt/filter/PageFilter.java]: not "Grants". Maybe a generic description will fit most of the (future) use cases. Something like 

```
Class used to filter the results received when calling an endpoint that supports Pagination.
```

File: src/main/java/com/auth0/client/mgmt/filter/RolesFilter.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
+package com.auth0.client.mgmt.filter;
+
+/**
+ * Class used to filter the results received when calling the Grants endpoint. Related to the {@link com.auth0.client.mgmt.RolesEntity} entity.
+ */
+public class RolesFilter extends BaseFilter {
```
Comment (lbalmaceda) [inline src/main/java/com/auth0/client/mgmt/filter/RolesFilter.java]: why reimplementing this? make it extend `PageFilter` and just add "withName"

File: README.md [hunk 9]
```diff
@@ -285,6 +285,7 @@ The Management API is divided into different entities. Each of them have the lis
 * **Jobs:** See [Docs](https://auth0.com/docs/api/management/v2#!/Jobs/get_jobs_by_id). Access the methods by calling `mgmt.jobs()`.
 * **Logs:** See [Docs](https://auth0.com/docs/api/management/v2#!/Logs/get_logs). Access the methods by calling `mgmt.logEvents()`. This endpoint supports pagination.
 * **Resource Servers:** See [Docs](https://auth0.com/docs/api/management/v2#!/Resource_Servers/get_resource_servers). Access the methods by calling `mgmt.resourceServers()`. This endpoint supports pagination.
+* **Roles:** See [Docs](https://auth0.com/docs/api/management/v2#!/Roles/get_roles). Access the methods by calling `mgmt.rules()`. This endpoint supports pagination.
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply README.md]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/RolesEntity.java [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+  }
+
+  /**
+   * Request all Roles created by this tenant that can be assigned to a given user or user group.
+   * A token with read:roles is needed
+   * See https://auth0.com/docs/api/management/v2#!/Roles/get_roles
+   *
+   * @return a Request to execute
+   */
+  public Request<List<Role>> list() {
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/RolesEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/RolesEntity.java [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+    String url = builder.build().toString();
+    CustomRequest<RolesPage> request = new CustomRequest<>(this.client, url, "GET", new TypeReference<RolesPage>() {});
+    request.addHeader("Authorization", "Bearer " + apiToken);
+    return request;
+  }
+
+  /**
+   * Get a single role created by this tenant that can be assigned to a given user or user group.
+   * A token with scope read:roles is needed.
+   * If you want the identities.access_token property to be included, you will also need the scope read:user_idp_tokens.
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/RolesEntity.java]: This has been implemented as suggested (and yup, that was copy-pasta, sorry).

File: src/main/java/com/auth0/client/mgmt/RolesEntity.java [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+
+    String url = builder.build().toString();
+    CustomRequest<Role> request = new CustomRequest<>(this.client, url, "GET", new TypeReference<Role>() {});
+    request.addHeader("Authorization", "Bearer " + apiToken);
+    return request;
+  }
+
+
+  /**
+   * Create a Role. A token with scope create:roles is needed.
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/RolesEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/RolesEntity.java [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+        .build()
+        .toString();
+    CustomRequest<Role> request = new CustomRequest<>(this.client, url, "POST", new TypeReference<Role>() {});
+    request.addHeader("Authorization", "Bearer " + apiToken);
+    request.setBody(role);
+    return request;
+  }
+
+  /**
+   * Delete an existing Role. A token with scope delete:roles is needed.
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/RolesEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/RolesEntity.java [hunk 14]
```diff
@@ -0,0 +1,336 @@
...
+        .addEncodedPathSegments(roleId)
+        .build()
+        .toString();
+    VoidRequest request = new VoidRequest(this.client, url, "DELETE");
+    request.addHeader("Authorization", "Bearer " + apiToken);
+    return request;
+  }
+
+  /**
+   * Update an existing Role. A token with scope update:roles is needed.
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/RolesEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/RolesEntity.java [hunk 15]
```diff
@@ -0,0 +1,336 @@
...
+  }
+
+  /**
+   * Lists the users that have been associated with a given role.
+   * See https://auth0.com/docs/api/management/v2#!/Roles/get_users
+   *
+   * @param roleId the role id
+   * @return a Request to execute
+   */
+  public Request<List<User>> listUsers(String roleId) {
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/RolesEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/RolesEntity.java [hunk 16]
```diff
@@ -0,0 +1,336 @@
...
+        .addEncodedPathSegments("users")
+        .build().toString();
+    CustomRequest<List<User>> request = new CustomRequest<>(this.client, url, "GET", new TypeReference<List<User>>() {});
+    request.addHeader("Authorization", "Bearer " + apiToken);
+    return request;
+  }
+
+  /**
+   * Lists the users that have been associated with a given role.
+   * See https://auth0.com/docs/api/management/v2#!/Roles/get_users
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/RolesEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/RolesEntity.java [hunk 17]
```diff
@@ -0,0 +1,336 @@
...
+      }
+    }
+    String url = builder.build().toString();
+    CustomRequest<UsersPage> request = new CustomRequest<>(this.client, url, "GET", new TypeReference<UsersPage>() {});
+    request.addHeader("Authorization", "Bearer " + apiToken);
+    return request;
+  }
+
+  /**
+   * Assign users to a role. A token with update:roles is needed.
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/RolesEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/RolesEntity.java [hunk 18]
```diff
@@ -0,0 +1,336 @@
...
+  }
+
+  /**
+   * Get the permissions associated to the role. A token with read:roles is needed.
+   * See https://auth0.com/docs/api/management/v2#!/Roles/get_permissions
+   *
+   * @param roleId the role id
+   * @return a Request to execute
+   */
+  public Request<List<Permission>> listPermissions(String roleId) {
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/RolesEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/RolesEntity.java [hunk 19]
```diff
@@ -0,0 +1,336 @@
...
+        .addEncodedPathSegments("api/v2/roles")
+        .addEncodedPathSegments(roleId)
+        .addEncodedPathSegments("permissions").build().toString();
+    CustomRequest<List<Permission>> request = new CustomRequest<>(this.client, url, "GET", new TypeReference<List<Permission>>() {});
+    request.addHeader("Authorization", "Bearer " + apiToken);
+    return request;
+  }
+
+  /**
+   * Get the permissions associated to the role. A token with read:roles is needed.
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/RolesEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/RolesEntity.java [hunk 20]
```diff
@@ -0,0 +1,336 @@
...
+   * See https://auth0.com/docs/api/management/v2#!/Roles/get_permissions
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/RolesEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/RolesEntity.java [hunk 21]
```diff
@@ -0,0 +1,336 @@
...
+    }
+    String url = builder.build().toString();
+    CustomRequest<UsersPage> request = new CustomRequest<>(this.client, url, "GET", new TypeReference<UsersPage>() {});
+    request.addHeader("Authorization", "Bearer " + apiToken);
+    return request;
+  }
+
+  /**
+   * Assign users to a role. A token with update:roles is needed.
+   * See https://auth0.com/docs/api/management/v2#!/Roles/post_users
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/RolesEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/RolesEntity.java [hunk 22]
```diff
@@ -0,0 +1,336 @@
...
+    }
+    String url = builder.build().toString();
+    CustomRequest<PermissionsPage> request = new CustomRequest<>(this.client, url, "GET", new TypeReference<PermissionsPage>() {});
+    request.addHeader("Authorization", "Bearer " + apiToken);
+    return request;
+  }
+
+  /**
+   * Un-associate permissions from a role. A token with update:roles is needed.
+   * See https://auth0.com/docs/api/management/v2#!/Roles/delete_permissions
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/RolesEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/RolesEntity.java [hunk 23]
```diff
@@ -0,0 +1,336 @@
...
+      }
+    }
+    String url = builder.build().toString();
+    CustomRequest<PermissionsPage> request = new CustomRequest<>(this.client, url, "GET", new TypeReference<PermissionsPage>() {});
+    request.addHeader("Authorization", "Bearer " + apiToken);
+    return request;
+  }
+
+  /**
+   * Un-associate permissions from a role. A token with update:roles is needed.
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/RolesEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/RolesEntity.java [hunk 24]
```diff
@@ -0,0 +1,336 @@
...
+    request.setBody(body);
+    request.addHeader("Authorization", "Bearer " + apiToken);
+    return request;
+  }
+
+  /**
+   * Associate permissions with a role. Only the `permission_name` and
+   * `resource_server_identifier` Permission attributes should be specified.
+   * A token with update:roles is needed.
+   * See https://auth0.com/docs/api/management/v2#!/Roles/post_permissions
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/RolesEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/UsersEntity.java [hunk 25]
```diff
@@ -350,4 +355,213 @@ public Request deleteMultifactorProvider(String userId, String provider) {
...
+
+
+    /**
+     * Get the permissions associated to the user. A token with read:users is needed.
+     * See https://auth0.com/docs/api/management/v2#!/Users/get_permissions
+     *
+     * @param userId the role id
+     * @return a Request to execute
+     */
+    public Request<List<Permission>> listPermissions(String userId) {
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/UsersEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/UsersEntity.java [hunk 26]
```diff
@@ -350,4 +355,213 @@ public Request deleteMultifactorProvider(String userId, String provider) {
...
+        String url = builder.build().toString();
+        CustomRequest<PermissionsPage> request = new CustomRequest<>(client, url, "GET", new TypeReference<PermissionsPage>() {
+        });
+        request.addHeader("Authorization", "Bearer " + apiToken);
+        return request;
+    }
+
+
+    /**
+     * Remove permissions from a user. A token with update:users is needed.
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/UsersEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/UsersEntity.java [hunk 27]
```diff
@@ -350,4 +355,213 @@ public Request deleteMultifactorProvider(String userId, String provider) {
...
+            .addPathSegments(userId)
+            .addPathSegments("permissions").build().toString();
+        CustomRequest<List<Permission>> request = new CustomRequest<>(client, url, "GET", new TypeReference<List<Permission>>() {
+        });
+        request.addHeader("Authorization", "Bearer " + apiToken);
+        return request;
+    }
+
+    /**
+     * Get the permissions associated to the user. A token with read:users is needed.
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/UsersEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/UsersEntity.java [hunk 28]
```diff
@@ -350,4 +355,213 @@ public Request deleteMultifactorProvider(String userId, String provider) {
...
+            .build()
+            .toString();
+        VoidRequest request = new VoidRequest(this.client, url, "DELETE");
+        request.setBody(body);
+        request.addHeader("Authorization", "Bearer " + apiToken);
+        return request;
+    }
+
+    /**
+     * Assign permissions to a user. A token with update:users is needed.
```
[the remark this answers is not in the document]
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/UsersEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/UsersEntity.java (hunk 1 above)
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/UsersEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/UsersEntity.java (hunk 2 above)
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/UsersEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/UsersEntity.java (hunk 3 above)
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/UsersEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/UsersEntity.java (hunk 4 above)
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/UsersEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/UsersEntity.java (hunk 5 above)
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/UsersEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/UsersEntity.java (hunk 6 above)
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/UsersEntity.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/filter/PageFilter.java (hunk 7 above)
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/filter/PageFilter.java]: This has been implemented as suggested.

File: src/main/java/com/auth0/client/mgmt/filter/RolesFilter.java (hunk 8 above)
Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/filter/RolesFilter.java]: Good call. This has been implemented as suggested.  Please note: I did not adjust any of the pre-existing Filters that also include pagination, as that was outside the scope of this particular PR.

Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/filter/RolesFilter.java]: Actually, I reverted this. Unless I introduce generics to the Filters, the builder / flow pattern doesn't quite operate the same way. For example, the below would no longer return a RolesFilter:

` RolesFilter filter = new RolesFilter().withTotals(true)`

**Comment by jakbutler** [comment]
I've implemented all but on of your requested changes; making the `GrantFilter` extend `PageFilter` did not make sense at this time, as it broke the builder pattern (we'd need to use generics or interfaces, which seemed out of scope for the intent of the requested change).

File: src/main/java/com/auth0/client/mgmt/filter/RolesFilter.java (hunk 8 above)
Comment (lbalmaceda) [reply src/main/java/com/auth0/client/mgmt/filter/RolesFilter.java]: @jakbutler There's no need to use generics. You can override the method on the subclass, call super and make it return "this" just like it's being done [here](https://github.com/auth0/auth0-java/blob/master/src/main/java/com/auth0/client/mgmt/filter/QueryFilter.java#L64-L67). The overriden method will define in it's signature the type being returned, which matches the subclass class.

Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/filter/RolesFilter.java]: @lbalmaceda - I'd love to see an example; if you attempt a similar change on your local instance, I think you'll understand my response better. I tried what you are describing and using a builder method of a superclass returns an instance of the superclass (not the extending / implementing class), which breaks existing tests unless we use interfaces or generics. On top of that, almost every other Filter class has pagination but does not (nor would, as part of this PR), extend `PageFilter`, so I'm not sure why this is a blocking issue.

Comment (jakbutler) [reply src/main/java/com/auth0/client/mgmt/filter/RolesFilter.java]: Not crazy about that approach, but I've implemented it as suggested and pushed.

**Pull request merged**